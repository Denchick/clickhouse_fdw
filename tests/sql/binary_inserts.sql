CREATE EXTENSION clickhouse_fdw;
CREATE SERVER loopback FOREIGN DATA WRAPPER clickhouse_fdw OPTIONS(dbname 'regression', driver 'binary');
CREATE USER MAPPING FOR CURRENT_USER SERVER loopback;

SELECT clickhousedb_raw_query('drop database if exists regression');
SELECT clickhousedb_raw_query('create database regression');
SELECT clickhousedb_raw_query('CREATE TABLE regression.ints (
    c1 Int8, c2 Int16, c3 Int32, c4 Int64
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');

SELECT clickhousedb_raw_query('CREATE TABLE regression.uints (
    c1 UInt8, c2 UInt16, c3 UInt32, c4 UInt64
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');

SELECT clickhousedb_raw_query('CREATE TABLE regression.floats (
    c1 Float32, c2 Float64
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');

SELECT clickhousedb_raw_query('CREATE TABLE regression.null_ints (
    c1 Int8, c2 Nullable(Int32)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');

IMPORT FOREIGN SCHEMA "a" FROM SERVER loopback INTO public;

/* ints */
INSERT INTO ints
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(1, 3) i;
SELECT * FROM ints ORDER BY c1;
INSERT INTO ints (c1, c4, c3, c2)
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(4, 6) i;
SELECT * FROM ints ORDER BY c1;

/* check dropping columns (that will change attnums) */
ALTER TABLE ints DROP COLUMN c1;
ALTER TABLE ints ADD COLUMN c1 SMALLINT;
INSERT INTO ints (c1, c2, c3, c4)
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(7, 8) i;
SELECT c1, c2, c3, c4 FROM ints ORDER BY c1;

/* check other number types */
INSERT INTO uints
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(1, 3) i;
SELECT * FROM uints ORDER BY c1;
INSERT INTO floats
	SELECT i * 1.1, i + 2.1 FROM generate_series(1, 3) i;
SELECT * FROM floats ORDER BY c1;

/* check nullable */
INSERT INTO null_ints SELECT i, case WHEN i % 2 = 0 THEN NULL ELSE i END FROM generate_series(1, 10) i;
SELECT * FROM null_ints ORDER BY c1;

DROP USER MAPPING FOR CURRENT_USER SERVER loopback;
DROP EXTENSION clickhouse_fdw CASCADE;
