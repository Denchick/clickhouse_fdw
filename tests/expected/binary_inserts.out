CREATE EXTENSION clickhouse_fdw;
SET datestyle = 'ISO';
CREATE SERVER loopback FOREIGN DATA WRAPPER clickhouse_fdw OPTIONS(dbname 'regression', driver 'binary');
CREATE USER MAPPING FOR CURRENT_USER SERVER loopback;
SELECT clickhousedb_raw_query('drop database if exists regression');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('create database regression');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE TABLE regression.ints (
    c1 Int8, c2 Int16, c3 Int32, c4 Int64
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE TABLE regression.uints (
    c1 UInt8, c2 UInt16, c3 UInt32, c4 UInt64
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE TABLE regression.floats (
    c1 Float32, c2 Float64
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE TABLE regression.null_ints (
    c1 Int8, c2 Nullable(Int32)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE TABLE regression.complex (
    c1 Int32, c2 Date, c3 DateTime, c4 String, c5 FixedString(10), c6 LowCardinality(String)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

SELECT clickhousedb_raw_query('CREATE TABLE regression.arrays (
    c1 Int32, c2 Array(Int32)
) ENGINE = MergeTree PARTITION BY c1 ORDER BY (c1);
');
 clickhousedb_raw_query 
------------------------
 
(1 row)

IMPORT FOREIGN SCHEMA "a" FROM SERVER loopback INTO public;
/* ints */
INSERT INTO ints
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(1, 3) i;
SELECT * FROM ints ORDER BY c1;
 c1 | c2 | c3 | c4 
----+----+----+----
  1 |  2 |  3 |  4
  2 |  3 |  4 |  5
  3 |  4 |  5 |  6
(3 rows)

INSERT INTO ints (c1, c4, c3, c2)
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(4, 6) i;
SELECT * FROM ints ORDER BY c1;
 c1 | c2 | c3 | c4 
----+----+----+----
  1 |  2 |  3 |  4
  2 |  3 |  4 |  5
  3 |  4 |  5 |  6
  4 |  7 |  6 |  5
  5 |  8 |  7 |  6
  6 |  9 |  8 |  7
(6 rows)

/* check dropping columns (that will change attnums) */
ALTER TABLE ints DROP COLUMN c1;
ALTER TABLE ints ADD COLUMN c1 SMALLINT;
INSERT INTO ints (c1, c2, c3, c4)
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(7, 8) i;
SELECT c1, c2, c3, c4 FROM ints ORDER BY c1;
 c1 | c2 | c3 | c4 
----+----+----+----
  1 |  2 |  3 |  4
  2 |  3 |  4 |  5
  3 |  4 |  5 |  6
  4 |  7 |  6 |  5
  5 |  8 |  7 |  6
  6 |  9 |  8 |  7
  7 |  8 |  9 | 10
  8 |  9 | 10 | 11
(8 rows)

/* check other number types */
INSERT INTO uints
	SELECT i, i + 1, i + 2, i+ 3 FROM generate_series(1, 3) i;
SELECT * FROM uints ORDER BY c1;
 c1 | c2 | c3 | c4 
----+----+----+----
  1 |  2 |  3 |  4
  2 |  3 |  4 |  5
  3 |  4 |  5 |  6
(3 rows)

INSERT INTO floats
	SELECT i * 1.1, i + 2.1 FROM generate_series(1, 3) i;
SELECT * FROM floats ORDER BY c1;
 c1  | c2  
-----+-----
 1.1 | 3.1
 2.2 | 4.1
 3.3 | 5.1
(3 rows)

/* check nullable */
INSERT INTO null_ints SELECT i, case WHEN i % 2 = 0 THEN NULL ELSE i END FROM generate_series(1, 10) i;
INSERT INTO null_ints(c1) SELECT i FROM generate_series(11, 13) i;
SELECT * FROM null_ints ORDER BY c1;
 c1 | c2 
----+----
  1 |  1
  2 |   
  3 |  3
  4 |   
  5 |  5
  6 |   
  7 |  7
  8 |   
  9 |  9
 10 |   
 11 |   
 12 |   
 13 |   
(13 rows)

SELECT * FROM null_ints ORDER BY c1;
 c1 | c2 
----+----
  1 |  1
  2 |   
  3 |  3
  4 |   
  5 |  5
  6 |   
  7 |  7
  8 |   
  9 |  9
 10 |   
 11 |   
 12 |   
 13 |   
(13 rows)

/* check dates and strings */
INSERT INTO complex VALUES
	(1, '1990-06-01', '1990-06-02 10:01:02', 't1', 'fix_t1', 'low1'),
	(2, '1990-06-02', '1990-06-03 10:01:02', 5, 'fix_t2', 'low2');
SELECT * FROM complex ORDER BY c1;
 c1 |     c2     |         c3          | c4 |   c5   |  c6  
----+------------+---------------------+----+--------+------
  1 | 1990-06-01 | 1990-06-02 10:01:02 | t1 | fix_t1 | low1
  2 | 1990-06-02 | 1990-06-03 10:01:02 | 5  | fix_t2 | low2
(2 rows)

/* check arrays */
INSERT INTO arrays VALUES
	(1, ARRAY[1,2]),
	(2, ARRAY[3,4,5]),
	(3, ARRAY[6,4]);
SELECT * FROM arrays ORDER BY c1;
 c1 |   c2    
----+---------
  1 | {1,2}
  2 | {3,4,5}
  3 | {6,4}
(3 rows)

DROP USER MAPPING FOR CURRENT_USER SERVER loopback;
DROP EXTENSION clickhouse_fdw CASCADE;
NOTICE:  drop cascades to 7 other objects
DETAIL:  drop cascades to server loopback
drop cascades to foreign table arrays
drop cascades to foreign table complex
drop cascades to foreign table floats
drop cascades to foreign table ints
drop cascades to foreign table null_ints
drop cascades to foreign table uints
